src/Psalm/Internal/Codebase/Functions.php

```diff
@@ -252,6 +256,100 @@ public function getFullyQualifiedFunctionNameFromString(string $function_name, S
+                break;
+            }
+        }
+
+        // We will search all functions for several patterns. This will
+        // be for all used namespaces, the global namespace and matched
+        // used functions.
+        $match_function_patterns = [
+            $stub . '*',
+        ];
+
+        if ($current_namespace_aliases) {
+            // As the stub still include the current namespace in the symbol,
+            // remove the current namespace to provide the function-only token
+            // and match against global functions and all imported namespaces.
+            // "Bar/foo" will become "foo".
+            if ($current_namespace_aliases->namespace
+                && strpos($stub, strtolower($current_namespace_aliases->namespace) . '\\') === 0
+            ) {
+                $stub = substr($stub, strlen($current_namespace_aliases->namespace) + 1);
+                $match_function_patterns[] = $stub . '*';
+            }
+
+            foreach ($current_namespace_aliases->functions as $alias_name => $function_name) {
+                if (strpos($alias_name, $stub) === 0) {
+                    try {
+                        $match_function_patterns[] = $function_name;
+                    } catch (\Exception $e) {
+                    }
+                }
+            }
+            foreach ($current_namespace_aliases->uses as $namespace_name) {
+                $match_function_patterns[] = $namespace_name . '\\' . $stub . '*';
+            }
+        }
+
+        $function_maps = [
+            $file_storage->functions,
+            $this->getAllStubbedFunctions(),
+            $this->reflection->getFunctions(),
+            $codebase->config->getPredefinedFunctions(),
+        ];
+
+        foreach ($function_maps as $function_map) {
+            foreach ($function_map as $function_name => $function) {
+                if (isset($matching_functions[$function_name])) {
+                    continue;
+                }
+                foreach ($match_function_patterns as $pattern) {
+                    if (substr($pattern, -1, 1) === '*') {
+                        if (strpos($function_name, rtrim($pattern, '*')) !== 0) {
+                            continue;
+                        }
+                    } elseif ($function_name !== $pattern) {
+                        continue;
+                    }
+                    if (is_bool($function)) {
+                        /** @var callable-string $function_name */
+                        if ($this->reflection->registerFunction($function_name) === false) {
+                            continue;
+                        }
+                        $function = $this->reflection->getFunctionStorage($function_name);
+                    }
+                    $matching_functions[$function_name] = $function;
+                }
+            }
+        }
+
+        return array_values($matching_functions);
```

Comment:
This can be simplified to

```suggestion
        $function_map = $file_storage->functions
            + $this->getAllStubbedFunctions()
            + $this->reflection->getFunctions()
            + $codebase->config->getPredefinedFunctions();

        foreach ($function_map as $function_name => $function) {
           foreach ($match_function_patterns as $pattern) {
                if (substr($pattern, -1, 1) === '*') {
                    if (strpos($function_name, rtrim($pattern, '*')) !== 0) {
                        continue;
                    }
                } elseif ($function_name !== $pattern) {
                    continue;
                }
                if (is_bool($function)) {
                    /** @var callable-string $function_name */
                    if ($this->reflection->registerFunction($function_name) === false) {
                        continue;
                    }
                    $function = $this->reflection->getFunctionStorage($function_name);
                }
                $matching_functions[$function_name] = $function;
            }
        }

        return $matching_functions;
```